draft.md

```diff
@@ -0,0 +1,101 @@
+# Draft IPLD Unixfs Spec
+
+## Basic Structure
+
+A Unixfs is either a file or a directory.
+The top level IPLD object is a CBOR map with at least two fields: `type` and `data`
+and maybe a few other such as a version string or a set of flags.
+The `type` field is either `file` or `dir`.
+
+## IPLD `file`
+
+If an IPLD file is a leaf its CID type is `raw` (0x55) and has no structure.
+Otherwise its CID type is `dag-cbor` (0x71).
+The `type` field is set to `file` and the `data` field is an CBOR array.
+Each element of the array is CBOR map with the following fields:
+
+  - `data`: link
+  - `size`: cumulative size of `data`
+  - `fsize`: (file size) cumulative size of the payload of `data`
+  
+The `fsize` field is omitted if the link is `raw` as it is the same value as size.
+
+## IPLD `dir`
+
+An IPLD `dir` represents a directory.
+Its CID type is `dag-cbor` (0x71).
+The `type` field set to `dir` and the data field is an CBOR map.
+The key of the map is a filename and is a CBOR text string encoded in UTF-8.
+The value of the map is another CBOR map with the following standard fields:
+
+  - `type`
+  - `exe`: CBOR boolean: executable bit
+  - `data`: normally a CBOR link, but can be other types depending on the value of the `type` field
+  - `size`: cumulative size of `data`
+  - `fsize`: (file size) cumulative size of the payload of `data`
+  - `fname`: CBOR byte string: original filename if it differs from the key
+
+And at least the following optional fields:
+
+  - `ro`: CBOR boolean: read only
+  - `mtime`: Modification time
+  - `attr`: CBOR Map: Extended attributes
+
+Additional fields may be defined.  All implementation specific or user
+defined fields should be stored under the `attr` field. 
+
+### Directory Types
+
+The type field is limited to a set of well defined values:
+
+  * _omitted_: regular file 
+  * `dir`: directory entry
+  * `special`: special file type (fifo, device, etc).
+     The `data` field is a CBOR Map with at least one field to describe the type.
+  * `symlink`: symbolic link.  The `data` field is the contents of the link.
+  * `other`: link to other IPLD object, links followed for GC and related operations 
+  * `unknown`: link to unknown objects, links not followed
+
+### Extended Attributes
+
+The extended attributes set is not well defined and can be used for vendor extensions and POSIX attributes that don't make sense on non-unix systems.
+Stripping this field MUST not change the meaning of the directory entry.
+These attributes SHOULD be passed along but do not have to be understood.
+
+Possible entries:
```

Comment:
Would "Extended Attributes" be a good place to optionally store explicit media type for problematic data types, as noted in https://github.com/ipfs/unixfs-v2/issues/11 ?